src/odfesqlodbc/es_communication.cpp

```diff
@@ -438,9 +473,65 @@ int ESCommunication::ExecDirect(const char* query, const char* fetch_size_) {
         return -1;
     }
     m_result_queue.push(std::unique_ptr< ESResult >(result));
+    if (!result->cursor.empty()) {
+        // If response has cursor, this thread will retrives more results pages asynchronously.
+        auto send_cursor_queries = std::async(std::launch::async, [&]() {
+            SendCursorQueries(result->cursor.c_str());
+        });
+    }
     return 0;
 }
 
+void ESCommunication::SendCursorQueries(const char* _cursor) {
+    if (_cursor == NULL)
+        return;
+    try {
+        std::string cursor(_cursor);
+        while (!cursor.empty() && (m_result_queue.size() < m_result_queue_capacity)) {
+            std::shared_ptr< Aws::Http::HttpResponse > response = nullptr;
+            IssueRequest(SQL_ENDPOINT_FORMAT_JDBC,
+                         Aws::Http::HttpMethod::HTTP_POST, ctype, "", response,
+                         "", cursor);
+            if (response == nullptr) {
+                m_error_message =
+                    "Failed to receive response from cursor. "
+                    "Received NULL response.";
+                LogMsg(ES_ERROR, m_error_message.c_str());
+                return;
+            }
+            ESResult* es_result = new ESResult;
```

Comment:
Why not create this as a unique_ptr up front? Right now there's a possible leak if you have an exception thrown before you push it onto the queue.